File: CoinChange2.py
class Solution:
    def coinChange(self, coins, amount):
        """
        :type coins: List[int]
        :type amount: int
        :rtype: int
        """
        def coinChange(value):
            if value == 0:
                return 0
            if value in dp:
                return dp[value]
            cur = 2147483647
            for coin in coins:
                if value >= coin:
                    temp = coinChange(value - coin)
                    if temp >= 0:
                        cur = min(cur, temp + 1)
            if cur != 2147483647:
                dp[value] = cur
            else:
                dp[value] = -1
            return dp[value]

        if amount == 0:
            return 0
        dp = {0: 0}
        return coinChange(amount)


if __name__ == '__main__':
    coins = [int(coin) for coin in input().split()]
    amount = int(input())
    solution = Solution()
    print(solution.coinChange(coins, amount))

File: CoinChange2.cpp
#include<iostream>
#include<vector>
#include<algorithm>

using namespace std;

class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        
        vector<int> dp(amount + 1, -1);
        dp[0] = 0;
        for(int i = 1; i <= amount; i++) {
        	for(int j = 0; j <coins.size(); j++) {
        		if(i - coins[j] >= 0) {
        			if(dp[i - coins[j]] != -1) {
        				if(dp[i] == -1) dp[i] = dp[i - coins[j]] + 1;
        				else dp[i] = min(dp[i], dp[i - coins[j]] + 1);
					}
				}
			}
		}
		
		return dp[amount];
        
    }
    
};

int main() {
	
	int n;
	cin>>n;
	
	vector<int> coins;
	for(int i = 0; i < n; i++) {
		int coin;
		cin>>coin;
		coins.push_back(coin);
	}
	int amount;
	cin>>amount;
	
	Solution *solution = new Solution();
	cout<<solution->coinChange(coins, amount);
	
	return 0;
	
}
